[WebGPU] buffer array lengths from automatically generated bind groups can overwrite actual bindings

mwyrzykowski opened the pull request:
#### 20c2476108c028f4cc0ea92661bb13e9880ca3db
<pre>
[WebGPU] buffer array lengths from automatically generated bind groups can overwrite actual bindings
<a href="https://bugs.webkit.org/show_bug.cgi?id=289006">https://bugs.webkit.org/show_bug.cgi?id=289006</a>
<a href="https://rdar.apple.com/145996237">rdar://145996237</a>

Reviewed by Tadeu Zagallo.

Array lengths were given binding indices which could overlap with binding
indices provided by the user, leading to incorret automatically generated
bind groups.

Address this by keeping array length bindings in their own range, specifically in the range
[lastValidBindingIndex + 1, lastValidBindingIndex + maxBindingsPerBindGroup]

* LayoutTests/fast/webgpu/nocrash/fuzz-289006-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-289006.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::isArrayLength):
Add helper.

(WebGPU::Device::createBindGroupLayout):
Ensure array lengths don&apos;t cause validation to fail.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::addPipelineLayouts):
Keep array lengths in their own address space.

Canonical link: <a href="https://commits.webkit.org/291643@main">https://commits.webkit.org/291643@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc84f5bca93c063a54c21a3025be4a685ae9adff

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93202 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12757 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2527 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98201 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43728 "Built successfully") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95252 "Build is in progress. Recent messages:") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21293 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71239 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28646 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96204 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9842 "Found 1 new test failure: fast/forms/ios/focus-input-via-button-no-scaling.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84344 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51571 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9536 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1981 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43042 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79821 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1987 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100230 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20254 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14882 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80263 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20506 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80261 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79572 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19844 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24119 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1470 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13369 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20238 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25415 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19925 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23385 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21666 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -882,8 +882,10 @@ static auto makeBindingLayout(WGPUBindGroupLayoutEntry& newEntry, auto& bindingM
+            uint32_t webBinding = entry.webBinding;
             if (auto& entryName = entry.name; entryName.length()) {
                 if (entryName.endsWith("_ArrayLength"_s)) {
+                    webBinding += limits().maxBindingsPerBindGroup;
```
Wouldn't that put it beyond the maximum number of bindings? Is that ok?

mwyrzykowski committed:
[WebGPU] buffer array lengths from automatically generated bind groups can overwrite actual bindings
https://bugs.webkit.org/show_bug.cgi?id=289006
rdar://145996237

Reviewed by Tadeu Zagallo.

Array lengths were given binding indices which could overlap with binding
indices provided by the user, leading to incorret automatically generated
bind groups.

Address this by keeping array length bindings in their own range, specifically in the range
[lastValidBindingIndex + 1, lastValidBindingIndex + maxBindingsPerBindGroup]

* LayoutTests/fast/webgpu/nocrash/fuzz-289006-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-289006.html: Added.
Add regression test.

* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::isArrayLength):
Add helper.

(WebGPU::Device::createBindGroupLayout):
Ensure array lengths don't cause validation to fail.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::addPipelineLayouts):
Keep array lengths in their own address space.

Canonical link: https://commits.webkit.org/291643@main
